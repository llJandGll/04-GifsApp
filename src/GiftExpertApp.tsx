import React, { useState } from 'react';
import { AddCategory, GiftGrid } from './components';

export const GiftExpertApp : React.FC = () => {
  const [categories, setCategories] = useState<Array<string>>([]);
  const [errorMessage, setErrorMessage] = useState("");
  

  const onNewCategory = ( newCategory : string ) : void  => {
    if ( categories.includes( newCategory )) {
      setErrorMessage('The category already exist');
      return;
    }
    setCategories( (category : Array<string>) => [ newCategory, ...category ]);
  }

  const handleError = ()   => {
    return {
      setErrorMessage : setErrorMessage,
      errorMessage : errorMessage
    }
  }

  return (
    <>
      <h1> GiftExpertApp </h1>
      <AddCategory 
        onNewCategory={ (value : string) => onNewCategory(value) }
        handleError={ handleError }
      />
      {
        categories.map( ( category ) => {
          return <GiftGrid key={ category } category ={ category } />
        })
      }
    </>
  )
}

