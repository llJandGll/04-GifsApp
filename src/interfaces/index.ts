export * from './AddCategoryProps';
export * from './CategoryProps';