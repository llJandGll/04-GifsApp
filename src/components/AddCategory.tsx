import { FormEvent, useState } from "react"
import { AddCategoryProps } from "../interfaces";
// import { Category } from "../../interfaces/CategoryInterface";


export const AddCategory : React.FC<AddCategoryProps> = ({ onNewCategory, handleError }) => {

  const [searchInputValue, setSearchInputValue] = useState("");
  const {errorMessage , setErrorMessage } = handleError();

  const onSubmit = ( event : FormEvent) : void  => {
    event.preventDefault();
    
    inputValidation( searchInputValue );

    onNewCategory( searchInputValue.toLocaleLowerCase() );

    setSearchInputValue('');

  }

  const inputValidation = ( inputValue : string ) : void => {
    if( inputValue.trim().length === 0 ) {
      setErrorMessage('the field is empty');
      return;
    }
  }
   
  const onAddCateryButton = ( ) : void  => {
   inputValidation( searchInputValue );
    onNewCategory( searchInputValue );
    setSearchInputValue('');
  }
  

  const onInputChange = ( { target } : FormEvent) : void  => {
    const { value } = target as HTMLInputElement;
    setSearchInputValue( value );
    
    setErrorMessage('');
  }
  
  
  return (
    <form onSubmit={ onSubmit }>
      <input 
       type="text" 
        placeholder="Buscar Gif"
        value={ searchInputValue }
        onChange={ onInputChange }
      />
      
    {
      errorMessage && <p>{ errorMessage }</p>
    }
       

    <button type="button"
      onClick={ () => onAddCateryButton() }
    >
      Add Category
    </button>
 
    </form>
  )
}
