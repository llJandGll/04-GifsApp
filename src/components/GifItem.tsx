import React from 'react'
import { Gif } from '../interfaces'

export const GifItem : React.FC<Gif> = ( { title, url }) => {
  return (
    <div className='card'>
    <img src={ url } alt={title} />
    <p>{title}</p>
  </div>
  )
}
