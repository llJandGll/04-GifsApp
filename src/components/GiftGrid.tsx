import React from "react";

import { CategoryProps, Gif } from "../interfaces";
import { GifItem } from "./GifItem";
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GiftGrid: React.FC<CategoryProps> = ({ category }) => {

  const { gifs , isLoading } = useFetchGifs( category );
 
  
  return (
    <>
      <h2>{ category }</h2>

      <div className="card-grid">
        
        {
          isLoading && <p>Loading...</p>
        }
        {
          gifs.map(( gif : Gif) => (
              <GifItem key={ gif.id } {...gif} />
            ))
        }
      </div>
    </>
  );
};
