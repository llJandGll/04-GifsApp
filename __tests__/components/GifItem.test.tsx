import { describe, expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import { GifItem } from '../../src/components';
import React from 'react';

describe('Test in GiftItem.tsx', () => { 
  
  const id : string = '1';
  const title : string = 'goku';
  const url : string = 'https://localhost/goku.gif';

  test('should show a snapshot', () => {
    const { container } = render(<GifItem id={ id } title={ title } url={ url } />);
    expect(container).toMatchSnapshot();
  });

  test('It should display the image with the indicated url and alt. ',() => {
    render(<GifItem id={ id } title={ title } url={ url } />);
    
    const { src , alt} = screen.getByRole('img') as HTMLImageElement;
    expect(src ).toBe(url);
    expect( alt ).toBe( title);

  });

  test('it should dispaly title in the component', () => { 
    
    render( <GifItem id={ id } title={ title } url={ url } />);
    const titleElement = screen.getByText(title);
    expect(titleElement).toBeTruthy();
  })
})