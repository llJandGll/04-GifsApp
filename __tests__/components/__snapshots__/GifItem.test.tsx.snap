// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test in GiftItem.tsx > should show a snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="goku"
      src="https://localhost/goku.gif"
    />
    <p>
      goku
    </p>
  </div>
</div>
`;
